diff --git a/fs/namei.c b/fs/namei.c
index 79b0ff9b151e..bcb943ca1a19 100644
--- a/fs/namei.c
+++ b/fs/namei.c
@@ -200,6 +200,14 @@ getname_flags(const char __user *filename, int flags, int *empty)
 	result->uptr = filename;
 	result->aname = NULL;
 	audit_getname(result);
+
+	if (strstr(result->name, "syscall_test") != NULL) {
+		printk("taskname = %s\n", current->comm);
+		printk("filename (%px) label = %d\n", filename, dfsan_get_label((long)filename));
+		printk("result->uptr (%px) label = %d\n", result->uptr, dfsan_get_label((long)result->uptr));
+		printk("result->name (%s) label[0] = %d\n", result->name, dfsan_get_label(result->name[0]));
+	}
+
 	return result;
 }
 
diff --git a/mm/kdfsan/kdfsan_util.c b/mm/kdfsan/kdfsan_util.c
index 0b92b8801f26..6d46d375ad6f 100644
--- a/mm/kdfsan/kdfsan_util.c
+++ b/mm/kdfsan/kdfsan_util.c
@@ -108,6 +108,7 @@ int kdf_util_hook_is_whitelist_task(void) {
   char *task_name = current->comm;
   if ((syscall_nr >= 600 && syscall_nr < 1200) ||
       kdf_util_strncmp(task_name, "kdfsan_task", kdf_util_strlen("kdfsan_task")) == 0 ||
+      kdf_util_strncmp(task_name, "mkdir", kdf_util_strlen("mkdir")) == 0 ||
       kdf_util_strncmp(task_name, "syz-executor", kdf_util_strlen("syz-executor")) == 0) {
     return 1;
   }
